src/typechecker.erl

```diff
@@ -1139,6 +1139,18 @@ new_type_var() ->
     I = erlang:unique_integer(),
     "_TyVar" ++ integer_to_list(I).
 
+bounded_type_list_to_type(TEnv, Types) ->
```

Comment:
I could have a go, but it'd be hard to test since we don't actually look at the generated constraints. For instance, with the current code you would expect this to fail:

```erlang
-spec id(A) -> A.
id(X) -> X.

-spec pair(foo, bar) -> {foo, bar}.
pair(X, Y) -> {id(X), id(Y)}.
```
since the same variable would be used for both calls to `id`. It doesn't and neither does,
```erlang
-spec pair(foo, bar) -> {foo, bar}.
pair(X, Y) -> {id(Y), id(X)}.
```
Creating fresh variables won't make a difference in these cases.

To do it properly we would also need to distinguish between flexible (freshly instantiated) variables and rigid (bound in the signature of the current function) variables.

Given all this, I think it makes more sense to save the instantiation for another PR (which I'd be happy to work on) where we try to address the problem more thoroughly.

Reply:
Ok, let's do this in another PR.